Workspace mapping

nury12n opened the pull request:
Creating a pull request for this in case there's time at the Scipy conference to address some of the bugs here.

This mostly a re-write of the mapping code to clean it up and make use of some of the workspace features. It won't be able to automatically merge with this branch because of the workspace stuff, but I wasn't sure if it would be better to make a pull request for the workspace or develop branch. If it makes more sense to have it as a pull request in one of those branches, I can close this one and move it to the other branch.

Changes
- Global min checking
  - After sampling phases, a DormantPhase for the phase with the highest sampled driving force will be creating to compute the driving force. 
  - An option exists to test n samples with the highest driving force, which can be useful for models with high internal DOFs where the sampling density may not capture the energy surface that well. 
  - The user also has the option to increase the sampling density through map kwargs
  - Same post-checks as before are used to detect degenerate composition sets to avoid some of the cases where the DLASCLS errors shows up
- Mapping and strategies
  - Mapper class is removed and the binplot/ternplot functions will now create an instance of a strategy instead
  - Iteration function is a bit cleaner, where it runs in the following order: iterate zpf line -> iterate node exits when current zpf line is finished -> iterate node queue when current exits are finished.
  - Automatic starting points are implemented in the strategies through the initialize function. This should make is easier for users to manually setup a strategy and map if needed.
  - Custom starting points will not need to specify which axis to map along since the strategy will find the starting direction during mapping. For step, binary and ternary strategies, all composition sets can be free and the strategy will determine which composition set to fix. This should make it easier for the user to add custom starting points if needed.
- Tieline Strategy
  - Tieline strategy is split into BinaryStrategy and TernaryStrategy. I found the difference in implementation to be different enough to warrant separate classes
  - Starting points are found by creating a StepStrategy along each axis and finding all the unique zpf lines from the StepStrategy. It will be good to check the performance of this and we may go back to the old method of generate starting points if it's a lot slower.
  - Starting directions and fixed composition sets are found using dot derivatives and comparing the slope of the ZPF line at each phase boundary. This removes the need for testing equilibrium in both axis directions which lead to some of the DLASCLS errors that we had before.
  - Ternary strategy
    - Starting points for ternaries will also step along the third axis of the ternary diagram (NOTE: this assumes species are pure elements and the user is mapping from 0-1 in composition space)
    - Direction testing in ternaries will test in the normal direction of the tie line. I forgot exactly what cases this helped in, but I remember there were cases where testing in the axis direction would fail due to the direction of the tieline and the phase boundary.
    - A second global min check is performed after finding a node to ensure a false global min is not found. If a false global min is detected, the incorrect node is replaced with a starting point
    - For nodes that are connected by two stoichiometric phases, starting a zpf line will fail (since it would not be able to step before finding a new global min), in which a starting point will be added in the normal of the node exit
- Isopleth strategy
  - Direction testing is based off dot derivatives similar to binary strategy
  - Removed some redundant calculations in exit finding
- Plotting
  - binplot and ternplot should retain the the same arguments as the old binplot/ternplot functions with the different plot args
  - binplot and ternplot also has the option to plot the endpoints of all the nodes
  - map_kwargs will allow the user to change some of the constants used in the strategies (global min sampling density, step size scaling, etc.)
  - Plotting functions in mapping.plotting (plot_step, plot_binary, plot_ternary, plot_isopleth) can support plotting different variables. By default, stepplot will plot axis variable vs. NP, binplot and isoplethplot will plot composition vs. temperature and ternplot will plot composition vs. composition. Currently, if the user wants to plot different variables, they will have to go through the mapping.plotting functions, with the exception of ternplot since it already had the x, y inputs.

TODOs
- There are still some DLASCLS errors. I have not gone around to finding the conditions that these occur in. I believe the three main situations are:
  - When stepping along a miscibility gap and reaching the peak of the miscibility gap, then there is a possibility that the global conditions will be very close to the composition with the peak is and the two composition sets will have almost the same composition. There's a check for whether the composition sets are two close to each other in composition, but it's like that it passes this check, then steps, then the new global condition leads to the same composition for the two phases.
  - Second order transitions? Ex. BCC_A2->BCC_B2 in Fe-Al and Al-Mn. I think you get a node where the order and disorder composition sets are at the same composition but different constituent site fractions
  - Some isopleths will show a DLASCLS error, but I hadn't looked more into it, but I think it might be a case where the temperature may go to +/- infinity on a near vertical phase boundary, although the axis should swap to temperature in those cases
- Global min check can fail to detect phases with high internal DOFs. Possible options:
  - Create DormantPhase for multiple phases with high driving force and then grab the phase with the highest driving force. There is a way to do it right now with the GLOBAL_MIN_NUM_CANDIDATE option, but it's a bit unintuitive where it checks the n samples, then creates a DormantPhase for all unique phases in the n samples, rather than testing n phases.
  - Keep all possible phases as a DormantPhase and update them per iteration. This can be slow though if there are a lot of phases
- Starting point generating using the StepStrategy can be a bit slow, especially for systems with a lot of phases or a lot of complex phases. And occasionally, I've seen it fail for ternaries and isopleths where no starting points can be found.
- Mapping can be sensitive to the ratio between the two axis variable step size, where the wrong direction will be proposed and the zpf line will not be able to be generated since the wrong direction will lead to an equilibrium failure. This is an unfortunate consequence of the new direction testing for binaries and ternaries where the direction is normalized to the step size. I seem to encounter this on systems with low melting points, where I will set the temperature step size too high. It could be possible to set a maximum step size based off the range of the axis variable, but I also don't want to change user inputs behind the scenes.
- Ternaries
  - If the fixed composition set leads to a miscibility gap, then it's possible that the node will be missed when we step in the free composition set and step directly over the node, and the fixed composition set will just move to the other side of the miscibility gap rather than being detected as a new global min
  - We should be able to remove the assumption of the user setting the two axis to (0, 1). This will mainly change how we do the starting points on the third axis. A example where a user may not want the entire composition range could be for a carbon containing system where the carbon content only goes up to a few percent.
- Mapping with a fixed chemical potential/activity/pressure variable?
- Isopleths
  - I have not tested isopleths where the two axes are composition
  - Mapping with linear combinations can lead to many failed equilibrium calculations.
  - A decent number of the isopleths I tested can miss an exit or two from a node
  - Axis limits should be affected by the additional fixed composition, where the axis variable will be limited between (0, 1-sum(fixed compositions)). Currently, the zpf line will stop when it passes the upper limit due to a failed equilibrium calculation (where sum(x) > 1), but it would be better to stop the zpf line before attempting equilibrium. 
- There could be an option to return the strategy from binplot/ternplot so that the user can create different plots, although, the user could also generate the map manually to create custom plots.
- Unit tests
  - There are unit tests for the checks that are performed after each zpf line iteration
  - I suppose for the strategies, we can pick a simple system for each and make sure it generates the correct zpf lines and nodes, testing only for the phases that are present in each zpf line.
  - I'll have to see if it's possible to set up a strategy at a specific point, such as just before a node, then it would be possible to have unit tests for specific conditions (this could be like finding a node, checking the node exits, reaching an axis limit, etc)

richardotis committed:
debugging force numerical result

richardotis committed:
debugging switch to energy level evaluation

richardotis committed:
WIP: attempt to fix numerical difference through forced type conversion

richardotis committed:
debugging see if disordered fcc has the same issue

richardotis committed:
WIP: Try not replacing the ordering contribution

richardotis committed:
WIP: Try unwrapping symbols only

richardotis committed:
BLD: deploy: Bump versions for wheel build/deploy

richardotis committed:
TST: test_calculate: Change test name

richardotis committed:
MAINT: test_equilibrium: Remove debug verbosity from equilibrium calls

richardotis committed:
FIX: Workspace: Use default_factory in TypedField if value is None

richardotis committed:
WIP: Workspace: plot: Better axis labeling for Quantities

richardotis committed:
ENH: units: Add support for 'atom' as unit, for J/mol<->eV/atom conversions

richardotis committed:
WIP: units: Add support for getitem syntax for setting display units of properties

richardotis committed:
WIP: fixes for models and T0 units

richardotis committed:
FIX: utils: unpack_phases with KeysView input

richardotis committed:
FIX: Workspace: phases attribute with order-disorder

richardotis committed:
FIX: Workspace: Use PhaseRecordFactory state_variables attribute

richardotis committed:
ENH: ChemicalPotential: Add units

richardotis committed:
FIX: ComponentsField: Do not keep instance-dependent state on the descriptor object

richardotis committed:
FIX: PhaseRecordFactory: reactive parameters attribute

richardotis committed:
FIX: ModelsField: Allow Model types to be specified

richardotis committed:
MAINT: PhaseRecordFactory/callables: Reorganize modules

richardotis committed:
FIX: property_framework: Switch default properties shape to scalars

richardotis committed:
ENH: variables: Move 'supported variables' to be dynamically computed

richardotis committed:
MAINT: base units -> implementation units

richardotis committed:
Workspace: do not add units to plot label when are there none

richardotis committed:
property_framework: Add units to property protocol

richardotis committed:
ENH: units: Add distinct 'fraction' units

richardotis committed:
TST: Add initial tests and fix comparison for properties

richardotis committed:
FIX/TST: Workspace: Add tzero test

richardotis committed:
WIP: property_framework: hashability of objects

richardotis committed:
TST: Add test_workspace_conditions_change_clear_result

richardotis committed:
TST: Add test_workspace_conditions_specify_units

richardotis committed:
FIX: minimizer: Types for index array

richardotis committed:
WIP: minimizer: Try fixing dtype of index arrays again

richardotis committed:
WIP try again

richardotis committed:
FIX: minimizer: dtype mismatch issues on some platforms

richardotis committed:
MAINT: equilibrium: coverage improvements

richardotis committed:
MAINT: Workspace: Rename dbf->database, comps->components

richardotis committed:
WIP: Workspace/variables: Cleanup for imperative conditions

richardotis committed:
WIP: Conditions with units

richardotis committed:
WIP: renderer support

richardotis committed:
MAINT/ENH: eqsolver: Make add_new_phases a cpdef function, for mapping

richardotis committed:
MAINT: equilibrium: py310 import compat

richardotis committed:
FIX: conditions/ternplot: compatibility

richardotis committed:
FIX: minimizer: Step size for small phase_amt changes

richardotis committed:
FIX: Workspace: Check for bare arrays without units in recompute()

richardotis committed:
ENH/DOC/WIP: Update some examples, first draft of new reference state support

richardotis committed:
FIX for ReferenceState example notebook

richardotis committed:
FIX: set_plot_renderer: incorrect variable name

richardotis committed:
DOC: Add metastability related examples

richardotis committed:
fixup

richardotis committed:
DOC: Move some examples into advanced section

richardotis committed:
DOC: Update more docs for Workspace

richardotis committed:
Merge branch 'develop' into dot_derivatives

richardotis committed:
DOC: Bump copyright year

richardotis committed:
DOC/MAINT: triangular: Remove deprecated cla() and update ternplot example

richardotis committed:
TST/DOC: Add more CPF tests and fix small error in metastability notebook

richardotis committed:
TST/FIX: Add CPF reference state tests and some small import fixes

richardotis committed:
MAINT: Workspace: Remove render() class function

richardotis committed:
TST: test_cpf_calculation: Numerical tests for properties

richardotis committed:
WIP: minimizer: Support for linear combinations of mole fractions as conditions

richardotis committed:
ENH: Create Conditions object for handling workspace conditions

richardotis committed:
WIP: Conditions: Small fixes, move get_system_spec back to Solver

richardotis committed:
Merge branch 'develop' into fix-large-branching

richardotis committed:
MAINT: Model: Remove unnecessary float coercion code

richardotis committed:
MAINT: Model: Try removing iteration count limit for Piecewise unwrapping

richardotis committed:
FIX: calculate: Still add some extra_points for very complex phases

richardotis committed:
FIX: minimizer: Gibbs phase rule compliance improvements

richardotis committed:
TST: test_equilibrium: Gibbs phase rule compliance (gh-468)

richardotis committed:
Merge branch 'develop' into fix-large-branching

richardotis committed:
Merge remote-tracking branch 'richardotis/fix-468' into dot_derivatives

richardotis committed:
Merge remote-tracking branch 'richardotis/fix-large-branching' into dot_derivatives

richardotis committed:
FIX/TST: tzero: Al-Zn database is not valid above 1700 Kelvin

richardotis committed:
FIX: calculate: Compute lingrid based on actual number of endmember pairs, not the max

richardotis committed:
Merge branch 'dot_derivatives' into add_lincomb_conds

richardotis committed:
WIP: checkpoint with minimal lincomb condition example working and all tests passing

richardotis committed:
FIX: Differentiability of props from metaproperties

richardotis committed:
FIX: PhaseRecordFactory: Add items() attribute

richardotis committed:
ENH: MassFraction: Support as condition and ComputableProperty

richardotis committed:
TST: Add mass fraction condition test

richardotis committed:
TST: Add linear combination of mole fractions test

richardotis committed:
Merge branch 'develop' into add_lincomb_conds

richardotis committed:
ENH: Conditions: add basic repr

richardotis committed:
Merge branch 'develop' into dot_derivatives

richardotis committed:
Merge branch 'develop' into dot_derivatives

richardotis committed:
Merge branch 'develop' into dot_derivatives

richardotis committed:
MAINT: Fix syntax of nogil-except functions for recent Cython versions

richardotis committed:
FIX: metaproperties: SymEngine truth value of relational error

richardotis committed:
WIP: calculate: Phase-local conditions support

richardotis committed:
WIP: LinearCombination: Add repr

richardotis committed:
FIX: CompositionSet: Use explicit indexing instead of slicing to work around contiguity bug

richardotis committed:
FIX: minimizer: Use MIN_DRIVING_FORCE_TO_ADD instead of hard-coded constant

richardotis committed:
Merge branch 'develop' into dot_derivatives

richardotis committed:
FIX: hitandrun: Add small numerical tolerance for calculating polytope face distances

richardotis committed:
WIP: hitandrun: Try to increase robustness

richardotis committed:
MAINT: calculate: Remove unused function

richardotis committed:
FIX: calculate/polytope: Handle case of zero-solutions in sampler

richardotis committed:
WIP: polytope: Performance improvements

richardotis committed:
WIP: polytope: Adjust numerical tolerancing to try to fix test failures on some platforms

richardotis committed:
WIP: Phase-local conditions support

richardotis committed:
ENH: conditions: Molar ratio condition support

richardotis committed:
ENH/TST: Workspace: Phase-local conditions support

richardotis committed:
MAINT: cartesian: Remove obsolete module

richardotis committed:
FIX/TST: equilibrium: Phases are correctly suspended when a subsystem is specified that causes the phase to have zero feasible points. Fixes gh-503

richardotis committed:
MAINT/DOC: Regenerate API documentation

richardotis committed:
fixup for missing RST file in docs

richardotis committed:
MAINT: calculate: Add shortcut path for no phase-local conditions

dependabot[bot] committed:
Build(deps): Bump pypa/cibuildwheel from 2.16.4 to 2.16.5 (#512)

Bumps [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) from 2.16.4 to 2.16.5.
- [Release notes](https://github.com/pypa/cibuildwheel/releases)
- [Changelog](https://github.com/pypa/cibuildwheel/blob/main/docs/changelog.md)
- [Commits](https://github.com/pypa/cibuildwheel/compare/v2.16.4...v2.16.5)

---
updated-dependencies:
- dependency-name: pypa/cibuildwheel
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bocklund committed:
Merge branch 'develop' into dot_derivatives

dependabot[bot] committed:
Build(deps): Bump pypa/gh-action-pypi-publish from 1.8.11 to 1.8.12 (#515)

Bumps [pypa/gh-action-pypi-publish](https://github.com/pypa/gh-action-pypi-publish) from 1.8.11 to 1.8.12.
- [Release notes](https://github.com/pypa/gh-action-pypi-publish/releases)
- [Commits](https://github.com/pypa/gh-action-pypi-publish/compare/v1.8.11...v1.8.12)

---
updated-dependencies:
- dependency-name: pypa/gh-action-pypi-publish
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Build(deps): Bump nwtgck/actions-netlify from 2.1 to 3.0 (#518)

Bumps [nwtgck/actions-netlify](https://github.com/nwtgck/actions-netlify) from 2.1 to 3.0.
- [Release notes](https://github.com/nwtgck/actions-netlify/releases)
- [Changelog](https://github.com/nwtgck/actions-netlify/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/nwtgck/actions-netlify/compare/v2.1...v3.0)

---
updated-dependencies:
- dependency-name: nwtgck/actions-netlify
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Build(deps): Bump pypa/gh-action-pypi-publish from 1.8.12 to 1.8.14 (#519)

Bumps [pypa/gh-action-pypi-publish](https://github.com/pypa/gh-action-pypi-publish) from 1.8.12 to 1.8.14.
- [Release notes](https://github.com/pypa/gh-action-pypi-publish/releases)
- [Commits](https://github.com/pypa/gh-action-pypi-publish/compare/v1.8.12...v1.8.14)

---
updated-dependencies:
- dependency-name: pypa/gh-action-pypi-publish
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Build(deps): Bump pypa/cibuildwheel from 2.16.5 to 2.17.0 (#520)

Bumps [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) from 2.16.5 to 2.17.0.
- [Release notes](https://github.com/pypa/cibuildwheel/releases)
- [Changelog](https://github.com/pypa/cibuildwheel/blob/main/docs/changelog.md)
- [Commits](https://github.com/pypa/cibuildwheel/compare/v2.16.5...v2.17.0)

---
updated-dependencies:
- dependency-name: pypa/cibuildwheel
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

nury12n committed:
FIX: Saving TDB with A*(B+C) expressions (#522)

* FIX: ignoring IF/THEN statements in TYPE_DEFINITION keywords

* Fix: Database saving retains parenthesis for nested Mul(Add()) arguements

* fix added test to handle potentially different behavior across OS

* minor changes in comments

* simplify Mul expr processing

---------

Co-authored-by: Ury <ury3@llnl.gov>

richardotis committed:
BLD: deploy.yaml: PyPI trusted publisher support (#523)

* BLD: deploy.yaml: PyPI trusted publisher support

* Add testpypi support

* fixup

* try adding whitespace

* try fixing names for steps

* does this work?

* try without environment

richardotis committed:
FIX: get_state_variables: Detect SystemMolesType (N) case

richardotis committed:
FIX: Workspace: Unwrap models for PhaseRecordFactory to allow pickling

richardotis committed:
FIX: Workspace: Type validation for models attribute

richardotis committed:
FIX: units: Set pycalphad's unit registry as pint application default, for pickling support

richardotis committed:
FIX: lower_convex_hull: Correct array indexing for linear combination conditions in multi-component systems

richardotis committed:
FIX: ReferenceState: DifferentiableComputableProperty support

dependabot[bot] committed:
Build(deps): Bump pypa/cibuildwheel from 2.17.0 to 2.19.0 (#527)

* Build(deps): Bump pypa/cibuildwheel from 2.17.0 to 2.19.0

Bumps [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) from 2.17.0 to 2.19.0.
- [Release notes](https://github.com/pypa/cibuildwheel/releases)
- [Changelog](https://github.com/pypa/cibuildwheel/blob/main/docs/changelog.md)
- [Commits](https://github.com/pypa/cibuildwheel/compare/v2.17.0...v2.19.0)

---
updated-dependencies:
- dependency-name: pypa/cibuildwheel
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* FIX/DOC: Remove triangular module from docs to workaround matplotlib issue 28234

* FIX/DOC: pin docutils version to workaround compat with nbsphinx-link

* FIX/DOC: Another sphinx workaround; disable config caching in doc-build

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Richard Otis <richard.otis@outlook.com>

richardotis committed:
FIX/BLD: NumPy 2.0 Support (#530)

ricardonpa committed:
Fix float exponents when writing a TDB file (TC ERROR DatabaseUtils) (#516)

* Fix float exponents when writing a TDB file (TC ERROR DatabaseUtils)

* Update previous exponent float fix + Fix line breaks causing TC reading function ERROR

* Generalized the line-break rules to accept a break before any number addition/subtraction that is not part of an EXP or LN/LOG operation. Updated the test in test_database.py to reflect changes made in tdb.py.

* Update #522 expressions and by removing extra spaces to resolve conflicts

* Add comment to clarify usage of regex expression.

* Update tdb.py to account for errors when trying to use int() on exponents with nested functions, i.e., non-numeric values.

* Updated `test_tc_printer_exp` to test if exponents are being written as integers

* Create `test_reflow_text_for_line_breaks` function to test if lines are being broken at correct positions

dependabot[bot] committed:
Build(deps): Bump codecov/codecov-action from 3 to 4 (#511)

* Build(deps): Bump codecov/codecov-action from 3 to 4

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 3 to 4.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/v3...v4)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

* Add Codecov token

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Richard Otis <richard.otis@outlook.com>

dependabot[bot] committed:
Build(deps): Bump pypa/cibuildwheel from 2.19.0 to 2.19.1 (#529)

Bumps [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) from 2.19.0 to 2.19.1.
- [Release notes](https://github.com/pypa/cibuildwheel/releases)
- [Changelog](https://github.com/pypa/cibuildwheel/blob/main/docs/changelog.md)
- [Commits](https://github.com/pypa/cibuildwheel/compare/v2.19.0...v2.19.1)

---
updated-dependencies:
- dependency-name: pypa/cibuildwheel
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Build(deps): Bump pypa/gh-action-pypi-publish from 1.8.14 to 1.9.0 (#528)

Bumps [pypa/gh-action-pypi-publish](https://github.com/pypa/gh-action-pypi-publish) from 1.8.14 to 1.9.0.
- [Release notes](https://github.com/pypa/gh-action-pypi-publish/releases)
- [Commits](https://github.com/pypa/gh-action-pypi-publish/compare/v1.8.14...v1.9.0)

---
updated-dependencies:
- dependency-name: pypa/gh-action-pypi-publish
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Commit:
mapping compatibility with pycalphad

Commit:
add mapping folder to setup.py and remove old files

Commit:
maximum numpy version < 2.0

Commit:
removed print statements and added condition in tests if node finding fails

richardotis committed:
FIX: polytope: Allow redundant constraints

richardotis committed:
Merge branch 'develop' into dot_derivatives

richardotis committed:
FIX/MAINT: numpy 2.0 fixes

richardotis committed:
FIX: eqsolver: Convert elements of conds_keys to str when accessing properties.coords. (Why did this previously work?!)

richardotis committed:
FIX/DOC: Model.shift_reference_state: Use absolute reference to prevent multiple cross-reference error in doc-build. In principle, this should go away eventually as the new ReferenceState becomes the default and the legacy implementation is deprecated.

richardotis committed:
FIX: polytope: Use original code path if constraint matrix is square

richardotis committed:
Revert "FIX: polytope: Use original code path if constraint matrix is square"

This reverts commit 52d4a179b949763775ec31c29fe7a31bd7eaf145.

richardotis committed:
WIP: polytope: Add temporary debugging statement for CI

richardotis committed:
WIP: polytope: More debugging statements

Commit:
Merge remote-tracking branch 'upstream/dot_derivatives' into workspace_mapping

Commit:
map kwargs in plotting

Commit:
minor fixes with plot labels

Commit:
undo setup changes

Commit:
numpy 2.0 support

richardotis committed:
WIP: polytope: Try higher precision

richardotis committed:
Revert "WIP: polytope: Try higher precision"

This reverts commit 757da63d2d2655b2e0121173095c9a4fc09beb48.

richardotis committed:
Revert "WIP: polytope: More debugging statements"

This reverts commit 7a6c2ddf8faa439081cfc05bb87b784c17f394ae.

richardotis committed:
Revert "WIP: polytope: Add temporary debugging statement for CI"

This reverts commit 20d8db9e2afec35b0800b8b9ac3975cf3dc3cd98.

richardotis committed:
WIP: polytope: Check arguments of D more carefully

richardotis committed:
WIP: polytope: Try setting rcond for pinv

Commit:
map kwargs for starting points and fix for plotting for zpflines+nodes with miscibility gap in binary mapping

richardotis committed:
WIP: polytope: Try setting rcond=-1

richardotis committed:
WIP: polytope: Try using scipy.linalg.pinv instead

richardotis committed:
WIP: polytope: Try clipping particular solution array

bocklund commented:
This is fantastic! Awesome work. I'll try to play with it some. 

For merging, I think once we get Workspace merged in (https://github.com/pycalphad/pycalphad/pull/432), then we can cherry-pick the commits you made on top of what will be the new `pycalphad/develop`. Then it probably makes sense to make a new PR directly from `nury12n/workspace_mapping` on to `pycalphad/develop` and we can close https://github.com/pycalphad/pycalphad/pull/517 as superseded by the new PR.

bocklund commented:
https://github.com/richardotis/pycalphad/compare/dot_derivatives...nury12n:pycalphad:workspace_mapping link to the GH diff against the current Workspace branch

bocklund commented:
A script I put together to plot all the binary phase diagrams in a directory:

```python
from pathlib import Path
import matplotlib.pyplot as plt
from pycalphad import Database, binplot, variables as v, DofError

def read_tdbs(path):
    database_name_objs = []
    for db_file in path.rglob("*.tdb"):
        try:
            db_obj = Database(str(db_file))
        except Exception as e:
            print(db_file, e)
            db_obj = None
        if db_obj is not None:
            database_name_objs.append((str(db_file), db_obj))
    return database_name_objs

def plot_phase_diagrams(database_name_objs, outdir, overwrite_existing=True, capture_output=False):
    outdir.mkdir(exist_ok=True, parents=True)
    num_databases = len(database_name_objs)
    for i, (db_filename, dbf) in enumerate(database_name_objs):
        db_name = str(Path(db_filename).name)
        # TODO: ideally put all the output for one database on the same line, but
        # probably not worthwhile if we don't capture stdout/stderr well to keep the
        # one-line summary intact.
        print(f"{db_name} ({i+1} / {num_databases})", flush=True)
        # TODO: doesn't consider the full path, so basedir/a/X-Y.tdb and basedir/b/X-Y.tdb will get flagged as duplicates
        outfile = outdir / f"pd-{db_name}.png"
        if outfile.exists() and not overwrite_existing:
            continue
        nonvacant_pure_elements = dbf.elements - {"/-", "VA"}
        if len(nonvacant_pure_elements) != 2:
            print(f"Got {len(nonvacant_pure_elements)} pure elements for {db_name}, skipping for now.")
            continue
        comps = sorted(nonvacant_pure_elements) + ["VA"]
        phases = list(dbf.phases.keys())
        conds = {v.P: 101325, v.T: (300, 3500, 20), v.N: 1, v.X(comps[1]): (0, 1, 0.05)}
        try:
            # TODO: could be nice to capture some timing info
            if capture_output:
                # currently not implemented
                # I tried to capture output with something like https://stackoverflow.com/questions/5136611/capture-stdout-from-a-script
                # so we could track DLASCL errors and suppress them automatically, but it seems to not capture that output, even though it's going to the stdout stream. Probably because it's from a c-extension.
                # Maybe something like this would work https://stackoverflow.com/a/29834357
                # for now, it's probably best to just redirect all stdout to a file.
                # then `cat pdout.txt | sed 's/ \*\*.*//g' | sed '/^$/d'` to show the output with removed errors
                ax = binplot_with_captured_output(db_name, dbf, comps, phases, conds)
            else:
                ax = binplot(dbf, comps, phases, conds)
        except (DofError, ZeroDivisionError) as e:
            print(f"Skipping {db_name} because it raised an Exception: {e}")
            continue
        ax.figure.savefig(outfile, bbox_inches="tight", dpi=150)
        plt.close(ax.figure)

if __name__ == "__main__":
    overwrite_existing = True  # If True, will always recompute a diagram. If False, skips (doesn't recompute) a phase diagram we already have. Useful for debugging.
    database_dir = Path("/path/to/databases")
    outdir = Path("binpds")

    db_name_objs = read_tdbs(database_dir)
    plot_phase_diagrams(db_name_objs, outdir, overwrite_existing=overwrite_existing)
```

Might be nice to move this over to the new PR when it gets made, just so it exists and stays somewhat searchable.

richardotis committed:
FIX: minimizer: delta_phase_amounts should be a size equal to the total number of compsets, not just the stable ones. Indexing is adjusted accordingly

Commit:
Merge remote-tracking branch 'upstream/dot_derivatives' into workspace_mapping

richardotis committed:
FIX: minimizer: Throw away infeasible lstsq solutions in _differential functions

Commit:
Merge remote-tracking branch 'upstream/dot_derivatives' into workspace_mapping

richardotis committed:
MAINT: minimizer: Add notes about lstsq destroying the input matrix

richardotis committed:
TST: Jansson derivatives: Consistency checking for derivatives relevant to phase diagram mapping. This test particularly focuses on the zero and 'undefined' cases.

richardotis committed:
TST/FIX: test_jansson_derivative_zero_and_undefined: Adjust numerical tolerance for assertion to deal with platform-dependent differences

richardotis committed:
ENH: Workspace: Add return_units kwarg to .get(), with False as default. Adjust tests as appropriate

richardotis committed:
ENH: Conditions: Assume display units for return, and pass back raw arrays

richardotis committed:
MAINT: fix numpy 2.0 deprecation warnings

richardotis committed:
MAINT/TST: Resolve/filter/catch all warnings, and turn all uncaught warnings into errors in pytest

richardotis committed:
MAINT: tdb: Remove reference to ast.Num (deprecated since py38)

richardotis committed:
MAINT: Resolve py312 deprecation for datetime.utcnow()

richardotis committed:
MAINT/BLD: Bump OS versions for wheel build/deploy workflow (#531)

richardotis committed:
FIX/BLD: Set fetch-depth to zero to guarantee version tags are picked up from SCM in sdist workflow (#533)

richardotis committed:
BLD/CI: Remove local version specifier for TestPyPI uploads (#534)

Removes local version specifier from CI builds to allow for upload of artifacts to TestPyPI

richardotis committed:
FIX: newline in deploy.yaml (#535)

richardotis committed:
ENH/BLD: Drop dedicated macOS universal wheel builds (#536)

richardotis committed:
MAINT/BLD: Drop Python 3.8 support (#537)

richardotis committed:
Merge branch 'develop' into dot_derivatives

richardotis committed:
FIX: wrap latex in raw strings to avoid SyntaxError

richardotis committed:
FIX: SiteFraction conditions are adjusted in the dilute limit, like MoleFraction is

richardotis committed:
MAINT: DormantPhase: Remove dot_derivative method for now (descope to rethink approach)

richardotis committed:
MAINT: fix deprecation warning for scalar conversion

richardotis committed:
TST: test_cpf_reference_state: Include Jansson derivatives in test

richardotis committed:
TST/FIX: test_site_fraction_conditions: test that dilute site fraction conditions work

richardotis committed:
FIX/TST: test_cpf_calculation: Mole fraction wildcards

richardotis committed:
TST: more tests to improve coverage of the CPF

richardotis committed:
MAINT: remove unused function

richardotis committed:
TST: test_dot_derivative_chempot_condition

richardotis committed:
TST/FIX: Correct test_property_repr

richardotis committed:
TST: test_miscibility_gap_cpf_specifier: Wildcard specifier tests for miscibility gaps

richardotis committed:
ENH/TST: test_condition_nonexistent_component

richardotis committed:
TST: test_mass_fraction_binary_dilute

richardotis committed:
ENH: Workspace: Remove kwargs from .get() and add .get_dict()

richardotis committed:
ENH/TST: Workspace.get_composition_sets() and fixes for zero-length condition arrays

richardotis committed:
ENH/MAINT: Change 'dot derivative' to 'Jansson derivative' in API, tests and docs

richardotis committed:
ENH: Workspace: Roll up property array dimensions to match conditions axis order

richardotis committed:
FIX: PandasRenderer: Unroll property array dimensions

richardotis committed:
FIX/MAINT: np.infty -> np.inf (numpy 2.0 compat)

richardotis committed:
DOC: Re-run example notebooks

Commit:
Merge remote-tracking branch 'upstream/dot_derivatives' into workspace_mapping

Commit:
sync to dot_derivatives and use JanssonDerivative api

Commit:
initial unit tests for mapping strategies

dependabot[bot] committed:
Build(deps): Bump pypa/cibuildwheel from 2.19.1 to 2.19.2 (#539)

Bumps [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) from 2.19.1 to 2.19.2.
- [Release notes](https://github.com/pypa/cibuildwheel/releases)
- [Changelog](https://github.com/pypa/cibuildwheel/blob/main/docs/changelog.md)
- [Commits](https://github.com/pypa/cibuildwheel/compare/v2.19.1...v2.19.2)

---
updated-dependencies:
- dependency-name: pypa/cibuildwheel
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

richardotis committed:
FIX/MAINT: np.infty -> np.inf (numpy 2.0 compat) (#538)

* FIX/MAINT: np.infty -> np.inf (numpy 2.0 compat)

* TST: test_plot: Add test to ensure binplot() will run

* MAINT: plot.triangular: Fix matplotlib deprecation for .cla()

* TST: test_plot: Change to Al-Mg to hit the codepath with the deprecation

bocklund committed:
ENH: Improve error message for ordering models having mismatched constituents (#541)

bocklund committed:
REL: 0.10.5

richardotis committed:
Merge branch 'develop' into dot_derivatives

richardotis committed:
FIX/TST: Computed Properties for invalid conditions should return NaN

richardotis committed:
FIX: calculate: Pure vacancy phases are correctly suspended (gh-503)

richardotis committed:
ENH: Workspace: .get() for single property returns without wrapping in a list

richardotis committed:
MAINT: minimizer: small lints

richardotis committed:
TST: test_lincomb_binary_condition_rhs_negative

richardotis committed:
WIP: minimizer: Debugging code

richardotis committed:
DOC: TernaryExamples: Add molar ratio and linear combination condition

richardotis committed:
WIP: minimizer: Try to avoid zero mass

richardotis committed:
WIP: minimizer: Guard lapack calls from NaN

richardotis committed:
WIP: minimizer: Remove unneeded system amount check

Commit:
Merge remote-tracking branch 'upstream/dot_derivatives' into workspace_mapping

Commit:
utility plot functions to separate data extraction from matplotlib interfacing

bocklund committed:
rename and whitespace cleanup

bocklund committed:
Import and whitespace fixes

bocklund committed:
Fix whitespace for comments

Commit:
updated some docstrings

Commit:
merge comflicts

Commit:
remove implicit vacancy addition to components

bocklund committed:
Update binplot docstring

bocklund committed:
Update ternary compat_api

Commit:
removed unused DOO+composition heuristic for sorting/fixing composition sets in starting points. Sorting/fixing composition sets is now determined from Jansson derivatives

Commit:
updated docstrings for plotting

Commit:
removed unused norm variable in ternary_strategy._sort_point

Commit:
minor docstring changes in strategies

Commit:
support for weight fraction plotting (mapping is still done in mole fraction)

Commit:
updated stepplot/isoplethplot docstrings to binplot/ternplot docstrings

Commit:
sort phases in strategy when plotting. V.X/v.X -> X()/W() conversions for axis labels

Commit:
chemical potential labeling and phase-local v.X/v.W plotting

richardotis committed:
ENH: Workspace API (#432)

- Adds a new `Workspace` object and associated imperative API. This object now underlies the `equilibrium` function, which remains for backwards compatibility. A `Workspace` object can be mutated after creation. The `Workspace.get()` function accepts `ComputableProperty` objects as input and returns arrays. Fixes #154 
- Adds a new `PhaseRecordFactory` object and associated connections to `PhaseRecord`, which enables deferred compilation of symbolically-defined model properties until they are called in the `PhaseRecord` API. The `PhaseRecord` object also gains `prop` and `prop_grad` functions for computing properties and property gradients of arbitrary symbolic attributes of `Model` instances.
- Creates a "Computable Property Framework" (CPF) for calculating properties of the system, individual phases, or some combination of these. It also provides a framework for defining custom properties.
- Adds support for Jansson derivatives (constrained total derivatives aka "dot" derivatives) of thermodynamic properties (fixes #261 )
- Adds support for T0 (tzero) calculation
- Adds support for driving force calculation, including nucleation driving forces, via the `IsolatedPhase` and `DormantPhase` meta-properties. Meta-properties are objects that know how to create new properties (as defined by the CPF). In this case, these meta-properties know how to start sub-calculations . The current way of finding starting points for these calculations is not ideal and prone to failure (basically reuses `CompositionSet` objects from the workspace equilibrium calculation), but it works well in the typical case.
- Adds physical units support via `pint`. All `ComputableProperty` objects have "implementation units" and "display units." Implementation units are what are assumed by the underlying numerical computation (which is done without units, for performance). The display units are what the user gets when `Workspace.get` is called. `ComputableProperty.__getitem__` allows the display units to be changed for individual calls to `Workspace.get`, or when setting `Workspace.conditions`.
- Also regarding physical units, a special unit conversion context is defined to allow conversions between per-mole properties and per-mass properties. There's room to add per-volume here as well, but that may not make it into this PR.
- Adds support for specifying linear combinations and ratios of mole fractions as a condition.
- Adds support for specifying weight fractions as a condition.
- Closes https://github.com/pycalphad/pycalphad/issues/133 through the multi-valued output from `Workspace.get` 
- Closes https://github.com/pycalphad/pycalphad/issues/163 because site fraction conditions can be given now
- Associated tests for these new features.
- `build_phase_records` is removed. Use the `PhaseRecord.prop_*` API in downstream functions instead.

Documentation
- Adds several examples for the Computable Property API, including driving force, T0, and isolated energy surfaces. A demo of export-to-DataFrame is also shown.
- Other examples have been updated to use the new APIs. Examples using the legacy APIs continue to be supported, but have been moved to a new "Advanced Examples" section. These sections will continue to be expanded and reorganized to accommodate our evolving best practices.

Commit:
minor docstring and whitespace changes in primitives and starting_points

bocklund committed:
merge develop post wks

bocklund committed:
remove-renderers

Commit:
Merge remote-tracking branch 'origin/workspace_mapping' into workspace_mapping

Commit:
merge recent workspace changes and fix compatibility

Commit:
fixed mapping unit tests

Commit:
added mapping to packages in setup.py

bocklund committed:
Revert bad merge changes

Commit:
mapping examples

Commit:
remove stepplot and isoplethplot

Commit:
minor changes

Commit:
Fix: ensure site fraction arrays in find_global_min_point are writeable before calling CompositionSet.update

Commit:
check binary and ternary TODOs

Commit:
simplify _check_axis_within_limit and remove _check_composition_within_limit

Commit:
docstrings on _check_axis_within_limit

Commit:
general _test_swap_axis method in MapStrategy

bocklund commented:
there’s an open PR in the pycalphad org repo
